torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1531,14 +1592,18 @@ void ProcessGroupNCCL::watchdogHandler() {
     for (auto it = workMetaList_.begin(); it != workMetaList_.end();
          /* no increment */) {
       auto& work = *it;
-      work.checkAndSetException();
+      // When terminateProcessGroup_ is true, communicators have already been
+      // aborted, So cannot check exception based on them. But watchdog needs to
+      // finish the check for the works that have already been enqueued to
+      // workMetaList_
+      if (!terminateProcessGroup_.load()) {
+        work.checkAndSetException();
+      }
       bool timedOut = work.checkTimeout();
 
       // If work hits an exception (either an error or timeout)
       if (work.exception()) {
         if (SHOULD_CLEAN_UP(asyncErrorHandling_)) {
-          // Abort work and corresponding communicators
-          work.abort();
```

Comment:
not sure i understand why you removed work.abort but kept abort.  and whether you intended to clean up this codepath in this PR or later.

Reply:
Yes, I wanted to clean up the work level abort eagerly, as work level abort does duplicate things as PG level abort, no need to call it twice. But if risky, I can leave it to later

Reply:
ideally i would leave it as a separate PR and take the time to explain the reasoning.  (not that it has to wait to later, could just be in this stack of PRs)

Reply:
Yes, the diff shown here is the 'outdated' version, the change has already been removed in the latest commit, marking as resolved